### PR TITLE
refactor BinaryMiddleware

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**

no

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

nothing

### CODE REVIEW

File: lib/serialization/BinaryMiddleware.js [hunk 1]
```diff
@@ -198,62 +243,42 @@ class BinaryMiddleware extends SerializerMiddleware {
...
-									i++;
-								}
-								break;
+						const size = NUMBERS_SIZES_BY_TYPE[type];
+						const writeNumber = getWriteNumberMethod(type, currentBuffer);
+
+						allocate(HEADER_SIZE + size * n);
+						writeU8(NUMBERS_HEADERS_BY_TYPE[type] | (n - 1));
+
+						while (n > 0) {
```
Comment (sokra) [inline lib/serialization/BinaryMiddleware.js]: I don't think this change is a good idea. I makes `writeNumber` and `readNumber` calls polymorphic which affects performance. Before that change v8 was able to inline and optimize these loops. Now I guess that this is no longer possible (not super sure).

Anyway I feel like this part of the code should be copied for each number type to make it obvious how performance is affected.

File: lib/serialization/BinaryMiddleware.js [hunk 2]
```diff
@@ -39,11 +39,11 @@ NopSection --> NopSectionHeaderByte
...
 
-F64NumbersSectionHeaderByte -> 0b001n_nnnn (n:length)
-I32NumbersSectionHeaderByte -> 0b010n_nnnn (n:length)
-I8NumbersSectionHeaderByte -> 0b011n_nnnn (n:length)
+F64NumbersSectionHeaderByte -> 0b001n_nnnn (n:count - 1)
+I32NumbersSectionHeaderByte -> 0b010n_nnnn (n:count - 1)
+I8NumbersSectionHeaderByte -> 0b011n_nnnn (n:count - 1)
 
-NullsSectionHeaderByte -> 0b0001_nnnn (n:length)
+NullsSectionHeaderByte -> 0b0001_nnnn (n:count - 1)
```
Comment (sokra) [inline lib/serialization/BinaryMiddleware.js]: :+1: good catch

File: lib/serialization/BinaryMiddleware.js (hunk 1 above)
Comment (vankop) [reply lib/serialization/BinaryMiddleware.js]: Should I write any explanation in code about this design?

**Comment by sokra** [comment]
Thanks

**Pull request merged**